From f88086abeb91161373346bcfff3d9c43dd61a183 Mon Sep 17 00:00:00 2001
From: Frank Earl <frank@earlconsult.com>
Date: Thu, 19 Jul 2018 13:29:55 -0400
Subject: Fix device tree entries for NanoPI Neo-Plus2.

There's a handful of gotchas that preclude things like temp management,
system ID, WiFi on, BT RF Kill, etc. that isn't IN the upstream that's
been kind of kludged in FriendlyElec's alleged "mainline" kernel that
is getting added in as a more modern, less hacked up item  This
hopefully fixes the Device Tree.  There's a few more patches to be
done to device drivers to make this fix work, so we'll add these in
afterwards.

diff --git a/arch/arm/boot/dts/sunxi-h3-h5.dtsi b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
index 7a83b1522..58e032a71 100644
--- a/arch/arm/boot/dts/sunxi-h3-h5.dtsi
+++ b/arch/arm/boot/dts/sunxi-h3-h5.dtsi
@@ -732,5 +732,25 @@
 				function = "s_cir_rx";
 			};
 		};
+
+	    sid: eeprom@01c14000 {
+		    compatible = "allwinner,sun8i-h3-sid";
+		    reg = <0x01c14000 0x400>;
+	    };
+
+		ths: ths@1c25000 {
+			compatible = "allwinner,sun8i-a33-ths";
+			reg = <0x01c25000 0x100>;
+			#thermal-sensor-cells = <0>;
+			#io-channel-cells = <0>;
+		};
+
+        thermal-zones {
+	        cpu_thermal: cpu_thermal {
+		        polling-delay-passive = <330>;
+		        polling-delay = <1000>;
+		        thermal-sensors = <&ths 0>;
+            };
+        };
 	};
 };
diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo-plus2.dts b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo-plus2.dts
index 1ed9f219d..45cd4169a 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo-plus2.dts
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5-nanopi-neo-plus2.dts
@@ -47,6 +47,7 @@
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
 #include <dt-bindings/pinctrl/sun4i-a10.h>
+#include <dt-bindings/thermal/thermal.h>
 
 / {
 	model = "FriendlyARM NanoPi NEO Plus2";
@@ -72,7 +73,8 @@
 
 		status {
 			label = "nanopi:red:status";
-			gpios = <&pio 0 20 GPIO_ACTIVE_HIGH>;
+			gpios = <&pio 0 10 GPIO_ACTIVE_HIGH>;
+			linux,default-trigger = "heartbeat";
 		};
 	};
 
@@ -97,6 +99,7 @@
 	vdd_cpux: gpio-regulator {
 		compatible = "regulator-gpio";
 		pinctrl-names = "default";
+		pinctrl-0 = <&vdd_cpux_r_npi>;
 		regulator-name = "vdd-cpux";
 		regulator-type = "voltage";
 		regulator-boot-on;
@@ -113,11 +116,146 @@
 	wifi_pwrseq: wifi_pwrseq {
 		compatible = "mmc-pwrseq-simple";
 		pinctrl-names = "default";
+		pinctrl-0 = <&wifi_en_npi>;
 		reset-gpios = <&r_pio 0 7 GPIO_ACTIVE_LOW>; /* PL7 */
 		post-power-on-delay-ms = <200>;
 	};
+
+	rfkill_bt {
+		compatible = "rfkill-gpio";
+		pinctrl-names = "default";
+		pinctrl-0 = <&bt_pwr_pin>;
+		reset-gpios = <&pio 6 13 GPIO_ACTIVE_HIGH>; /* PG13 */
+		clocks = <&osc32k>;
+		clock-frequency = <32768>;
+		rfkill-name = "sunxi-bt";
+		rfkill-type = "bluetooth";
+	};
+};
+
+&cpu0 {
+        #cooling-cells = <2>;
+        cooling-min-level = <0>;
+        cooling-max-level = <7>;
+        cpu-supply = <&vdd_cpux>;
+        operating-points-v2 = <&cpu0_opp_table>;
+};
+
+&cpu1 {
+        #cooling-cells = <2>;
+        cooling-min-level = <0>;
+        cooling-max-level = <7>;
+        cpu-supply = <&vdd_cpux>;
+        operating-points-v2 = <&cpu0_opp_table>;
+};
+
+&cpu2 {
+        #cooling-cells = <2>;
+        cooling-min-level = <0>;
+        cooling-max-level = <7>;
+        cpu-supply = <&vdd_cpux>;
+        operating-points-v2 = <&cpu0_opp_table>;
+};
+
+&cpu3 {
+        #cooling-cells = <2>;
+        cooling-min-level = <0>;
+        cooling-max-level = <7>;
+        cpu-supply = <&vdd_cpux>;
+        operating-points-v2 = <&cpu0_opp_table>;
+};
+
+&cpu0_opp_table {
+    opp-120000000 {
+        opp-hz = /bits/ 64 <120000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-240000000 {
+        opp-hz = /bits/ 64 <240000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-312000000 {
+        opp-hz = /bits/ 64 <312000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-480000000 {
+        opp-hz = /bits/ 64 <480000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-624000000 {
+        opp-hz = /bits/ 64 <624000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-816000000 {
+        opp-hz = /bits/ 64 <816000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-1008000000 {
+        opp-hz = /bits/ 64 <1008000000>;
+        opp-microvolt = <1100000>;
+        clock-latency-ns = <244144>;
+    };
+
+    opp-1248000000 {
+        opp-hz = /bits/ 64 <1248000000>;
+        opp-microvolt = <1300000>;
+        clock-latency-ns = <244144>;
+    };
+};
+
+&cpu_thermal {
+    trips {
+        cpu_warm: cpu_warm {
+	        temperature = <65000>;
+	        hysteresis = <2000>;
+	        type = "passive";
+        };
+        cpu_hot: cpu_hot {
+	        temperature = <75000>;
+	        hysteresis = <2000>;
+	        type = "passive";
+        };
+        cpu_very_hot: cpu_very_hot {
+	        temperature = <90000>;
+	        hysteresis = <2000>;
+	        type = "passive";
+        };
+        cpu_crit: cpu_crit {
+	        temperature = <105000>;
+	        hysteresis = <2000>;
+	        type = "critical";
+        };
+    };
+
+    cooling-maps {
+        cpu_warm_limit_cpu {
+	        trip = <&cpu_warm>;
+	        cooling-device = <&cpu0 THERMAL_NO_LIMIT 1>;
+        };
+        cpu_hot_limit_cpu {
+	        trip = <&cpu_hot>;
+	        cooling-device = <&cpu0 2 3>;
+        };
+        cpu_very_hot_limit_cpu {
+	        trip = <&cpu_very_hot>;
+	        cooling-device = <&cpu0 5 THERMAL_NO_LIMIT>;
+        };
+    };
 };
 
+
 &codec {
 	allwinner,audio-routing =
 		"Line Out", "LINEOUT",
@@ -130,6 +268,14 @@
 	status = "okay";
 };
 
+&ehci1 {
+	status = "okay";
+};
+
+&ehci2 {
+	status = "okay";
+};
+
 &ehci3 {
 	status = "okay";
 };
@@ -189,6 +335,14 @@
 	status = "okay";
 };
 
+&ohci1 {
+	status = "okay";
+};
+
+&ohci2 {
+	status = "okay";
+};
+
 &ohci3 {
 	status = "okay";
 };
@@ -199,6 +353,51 @@
 	status = "okay";
 };
 
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pins>, <&uart1_rts_cts_pins>;
+	status = "okay";
+};
+
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pins>, <&uart2_rts_cts_pins>;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pins>, <&uart3_rts_cts_pins>;
+	status = "okay";
+};
+
+&pio {
+	bt_pwr_pin: bt_pwr_pin@0 {
+		pins = "PG13";
+		function = "gpio_out";
+	};
+
+	uart2_rts_cts_pins: uart2_rts_cts {
+		pins = "PA2", "PA3";
+		function = "uart2";
+	};
+};
+
+&r_pio {
+	vdd_cpux_r_npi: regulator_pins@0 {
+		pins = "PL6";
+		function = "gpio_out";
+		drive = <SUN4I_PINCTRL_10_MA>;
+		pull = <SUN4I_PINCTRL_NO_PULL>;
+	};
+
+	wifi_en_npi: wifi_en_pin {
+		pins = "PL7";
+		function = "gpio_out";
+    };
+};
+
+
 &usb_otg {
 	dr_mode = "host";
 	status = "okay";
@@ -208,3 +407,5 @@
 	/* USB Type-A ports' VBUS is always on */
 	status = "okay";
 };
+
+
diff --git a/arch/arm64/boot/dts/allwinner/sun50i-h5.dtsi b/arch/arm64/boot/dts/allwinner/sun50i-h5.dtsi
index e237c05cf..82ce30dd6 100644
--- a/arch/arm64/boot/dts/allwinner/sun50i-h5.dtsi
+++ b/arch/arm64/boot/dts/allwinner/sun50i-h5.dtsi
@@ -47,28 +47,29 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		cpu@0 {
+		cpu0: cpu@0 {
 			compatible = "arm,cortex-a53", "arm,armv8";
 			device_type = "cpu";
 			reg = <0>;
 			enable-method = "psci";
+			clocks = <&ccu CLK_CPUX>;
 		};
 
-		cpu@1 {
+		cpu1: cpu@1 {
 			compatible = "arm,cortex-a53", "arm,armv8";
 			device_type = "cpu";
 			reg = <1>;
 			enable-method = "psci";
 		};
 
-		cpu@2 {
+		cpu2: cpu@2 {
 			compatible = "arm,cortex-a53", "arm,armv8";
 			device_type = "cpu";
 			reg = <2>;
 			enable-method = "psci";
 		};
 
-		cpu@3 {
+		cpu3: cpu@3 {
 			compatible = "arm,cortex-a53", "arm,armv8";
 			device_type = "cpu";
 			reg = <3>;
@@ -92,6 +93,12 @@
 			     <GIC_PPI 10
 				(GIC_CPU_MASK_SIMPLE(4) | IRQ_TYPE_LEVEL_LOW)>;
 	};
+
+	cpu0_opp_table: opp_table0 {
+	    compatible = "operating-points-v2";
+	    opp-shared;
+	};
+
 };
 
 &ccu {
